src/main/java/com/google/sps/model/GeocodingClientImpl.java

```diff
@@ -0,0 +1,52 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.model;
+
+import com.google.maps.GeoApiContext;
+import com.google.maps.GeocodingApi;
+import com.google.maps.GeocodingApiRequest;
+import com.google.maps.errors.ApiException;
+import com.google.maps.model.GeocodingResult;
+import com.google.sps.exceptions.GeocodingException;
+import java.io.IOException;
+
+/** Handles GET requests to the Google Geocoding API */
+public class GeocodingClientImpl implements GeocodingClient {
+  private GeocodingApiRequest geocodingService;
+
+  private GeocodingClientImpl(GeocodingApiRequest service) {
+    geocodingService = service;
+  }
+
+  /** Factory to create a GeocodingClientImpl instance with given API key */
+  public static class Factory implements GeocodingClientFactory {
+    @Override
+    public GeocodingClient getGeocodingClient(String apiKey) {
+      GeoApiContext context = new GeoApiContext.Builder().apiKey(apiKey).build();
+      GeocodingApiRequest service = GeocodingApi.newRequest(context);
+      return new GeocodingClientImpl(service);
+    }
+  }
+
+  @Override
+  public GeocodingResult getGeocodingResult(String address) throws GeocodingException {
+    try {
+      GeocodingResult[] response = geocodingService.address(address).await();
+      return response[0];
```

Comment:
Why would you have multiple results? What does it mean to have no results at all? This should be documented in the method documentation in `GeocodingClient` interface.

Would it make more sense to return the entire list to the servlet, and let it decide what these situations mean?

Reply:
The Geocoding API could return multiple results in case of partial matches and in that case the first match is the best match. No results would still return a GeocodingResult array (should probably handle this just in case regardless of the assumptions). This is under the assumption that origin and destination provided by the user are valid addresses (the only inputs that do not go through other sorts of validation) because the Maps API does not handle address validation. I agree that this should be documented as well but I don't think there's going to be any use of the rest of the results in the event of partial matches. Let me know what your thoughts are!

Reply:
I agree regarding explicitly checking the length of `response` and returning a more helpful error message, like actually saying there were no matches.

Reply:
I agree, I'll be sure to throw an error message on any index related access to `response` in https://github.com/googleinterns/blueprint-step-2020/pull/144 before merging because this is part here is the only one that accesses response in the PR and it is replaced to return a list in https://github.com/googleinterns/blueprint-step-2020/pull/144. I would like to keep the list because some of the methods iterate through the list of results to find the "best" one.

Reply:
`RESOLVED`